partner_email_check/models/res_partner.py

```diff
@@ -9,30 +9,84 @@
 _logger = logging.getLogger(__name__)
 
 try:
-    from validate_email import validate_email
+    from email_validator import validate_email, EmailNotValidError
 except ImportError:
-    _logger.debug('Cannot import "validate_email".')
+    _logger.debug('Cannot import "email_validator".')
 
-    def validate_email(email):
-        _logger.warning(
-            'Can not validate email, '
-            'python dependency required "validate_email"')
-        return True
+    validate_email = None
 
 
 class ResPartner(models.Model):
     _inherit = 'res.partner'
 
-    @api.constrains('email')
-    def constrains_email(self):
-        for rec in self.filtered("email"):
-            self.email_check(rec.email)
-
     @api.model
     def email_check(self, emails):
-        for email in emails.split(','):
-            if not validate_email(email):
-                raise ValidationError(
-                    _("%s is an invalid email") % email.strip()
-                )
-        return True
+        return ','.join(self._normalize_email(email.strip())
+                        for email in emails.split(','))
+
+    def _normalize_email(self, email):
+        if validate_email is None:
+            _logger.warning(
+                'Can not validate email, '
+                'python dependency required "email_validator"')
+            return email
+
+        try:
+            result = validate_email(
+                email,
+                check_deliverability=self._should_check_deliverability(),
+            )
+        except EmailNotValidError as e:
+            raise ValidationError(
+                _("%s is an invalid email: %s") % (email.strip(), e.message)
+            )
+        return result['local'].lower() + '@' + result['domain_i18n']
+
+    def _should_filter_duplicates(self):
+        conf = self.env['ir.config_parameter'].get_param(
+            'partner_email_check_filter_duplicates', 'False'
+        )
+        return conf == 'True'
+
+    def _should_check_deliverability(self):
+        conf = self.env['ir.config_parameter'].get_param(
+            'partner_email_check_check_deliverability', 'False'
+        )
+        return conf == 'True'
+
+    def _validate_no_duplicates(self, email):
+        is_existing_record = bool(self)
+        if is_existing_record:
+            self.ensure_one()
```

Comment:
this will break when some code does a write on multiple records at once, something like `$model.browse([1,2,3]).write({'email': 'a@b.c'})`.

What's your reasoning not to implement the whole thing as an ORM constraint? Then you also don't need to care about corner cases, because you know this record exists in the current transaction